Random variables can be used as given_condition

czgdp1807 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
Fixes #12237 
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed
Probability can now work if random variables are
passed as given_condition. A regression test
has been added

#### Other comments
I came to know about this issue from [here](https://gitter.im/sympy/sympy).
Since no PR referenced the above issue, I have made an attempt to fix the same.

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* stats
  * Random variable can be used as given_condition
<!-- END RELEASE NOTES -->

gdp1807 committed:
Random variables can be used as given_condition

Probability can now work if random variables are
passed as given_condition. A regression test
has been added

czgdp1807 commented:
@Upabjojr Can you please give your view on the approach I have used to fix the issue?
If you expected something else, then please guide me on that. :-)

oscarbenjamin commented:
It looks like there are some test failures as a result of this change...

czgdp1807 commented:
Yeah, I will look into those failures.

czgdp1807 commented:
I did some research on the internet and tried some debugging with the following code.
```
from sympy import *
from sympy.stats import *
from sympy.abc import x,y,z

X = Normal('X', 0, 1)
Y = Normal('Y', 0, 1)
B = density(X + Y, Y)(z)
```
I arrived at the following conclusions and results,
1. The `given` method in `rv.py`, converts the random variable Y into a symbol. I think this should not happen and this PR fixes the issue.
2. The random variable gives `X + Y` is independent from `Y` and hence, the output should have been according to the result given on [this wiki page](https://en.wikipedia.org/wiki/Sum_of_normally_distributed_random_variables).
3. Some improvements can also be made, like, `density(X + Y)(z)` gives unnecessary complex results, `(-erfc(z/2) + 2)*exp(-z**2/4)/(4*sqrt(pi)) + exp(-z**2/4)*erfc(z/2)/(4*sqrt(pi))`. However, the answer is, `Normal(0, sqrt(2))`.

Waiting for a reply and correct me if I am wrong.

czgdp1807 committed:
modified examples

oscarbenjamin commented:
If you look at the test results (details link below) you can see that many tests have failed because of the changes you made here. Have you seen those?

czgdp1807 commented:
Yes. I have seen those. I wanted some reviews on my previous comment. Hence, I am waiting to resolve those failures.

jksuom commented on sympy/stats/rv.py:
```diff
@@ -586,14 +586,16 @@ def given(expr, condition=None, **kwargs):
-    ------------------
-             ____
-         2*\/ pi
+                           2       2
```
This expression does not look like a density. I think that a conditional density, given `Y`, should also depend on `Y`.

czgdp1807 replied:
Yes you are right. I will resolve this. I did a mistake in analyzing.

jksuom replied:
I agree that the expression should be simplified.

czgdp1807 replied:
I am thinking to use MGF in the case of normal distributions. Since, Normal distributions are completely determined by their second moment. We can generalize it for X1 + X2 + ... + Xn.

czgdp1807 replied:
I discussed with @Upabjojr and agree with him that ad-hoc techniques like, using MGF won't be a good idea for computer algebra systems. So, @jksuom if you have any generalised technique to suggest to simplify the result then it would be great.

Upabjojr replied:
Simplification should not be handled by the _stats_ module. The _stats_ module should just compute things correctly. There is a separate _simplify_ module for simplifications.

czgdp1807 replied:
Yes, I agree for that too, it will make the code slower also as you suggested. `DiracDelta(Z - f(X, Y))` is fine for correctness. Maybe using some pre-existing results is the best approach, I can think as of now.

oscarbenjamin commented:
I don't know the stats part of the codebase very well at all. If I understand correctly this PR is stalled because you are waiting for a reviewer to respond to this comment:
[https://github.com/sympy/sympy/pull/15836#issuecomment-462041412](https://github.com/sympy/sympy/pull/15836#issuecomment-462041412)

czgdp1807 commented:
@oscarbenjamin I have got the reviews, I will proceed as soon as I am free of my exams. :-)

Upabjojr commented on sympy/stats/rv.py:
```diff
@@ -725,7 +728,7 @@ def probability(condition, given_condition=None, numsamples=None,
 
     if given_condition is not None and \
-            not isinstance(given_condition, (Relational, Boolean)):
+            not isinstance(given_condition, (Relational, Boolean, RandomSymbol)):
```
This change would require a lot of testing.

czgdp1807 replied:
Agreed, though, is this change valid or the current code in master works fine?
I think for using `RandomSymbol` as `given_condition`, this should be there. Though, expressions like, `X+Y` will create a problem since, they are an object of `Add` class. 
I hope that I am not misinterpreting the requirement of the issue.

Upabjojr commented on sympy/stats/tests/test_rv.py:
```diff
@@ -227,3 +227,7 @@ def test_issue_8129():
+
+def test_issue_12237():
+    X = Normal('X', 0, 1)
+    assert P(X > 0, X) == 1/2
```
isn't this supposed to be a function?

Upabjojr replied:
I mean, I would expect `P(A, B)` to create a `RandomSymbol` or similar.

czgdp1807 replied:
I cannot get your question. 
This test is passing, so any other issues with this? 
 Thanks for all of the reviews and discussion. :-)

czgdp1807 replied:
Oh I see, @Upabjojr  this wasn't visible to me a few minutes back, got it.
> I mean, I would expect `P(A, B)` to create a `RandomSymbol` or similar.
So, is this the accurate requirement of the issue as well?

Upabjojr replied:
Well, supporting random variables in the condition of `P` is much more complicated than your solution in this PR. `P(X>0, X)` is not a number, it's a random variable itself. This result is definitely wrong.

Upabjojr replied:
@czgdp1807 can you determine which random variable is `P(X>0, X)` equivalent to, in this case?

czgdp1807 replied:
If `Z = P(X>0, X=z)` and `Z` is a random variable then, according to my point of view, `Z` follows the distribution such that when, `z > 0`, the probability is 1 and when `z <= 0` then probability is 0.
A short explanation:
If z > 0, then we are given that X has some positive value, so `P(X > 0, X = z)` will be 1 and if z < 0, then we are given that X > 0 is not possible, so,`P(X > 0, X = z)` will be 0. We can also say that it's [discrete unit step function](https://en.wikipedia.org/wiki/Heaviside_step_function#Discrete_form).
Please correct if I am wrong.

Upabjojr replied:
I see it like this. A random variable maps events to real numbers. The density function it associates to real numbers it the probability of the events. So `P(X>0|X)` maps event `E` to `P(X>0|X = X(E))`. If `X(E) > 0`, which happens 1/2 of the times, then the value of `P(X>0)` is 1, otherwise it's 0.

I would say that `P(X>0|X)` is a Bernoulli random variable with `p=1/2` when `X` is normal centered at the origin.

Upabjojr replied:
I think your code is returning a completely wrong result.

czgdp1807 replied:
Yes @Upabjojr your opinion is right. Infact, `P(Z = 1) = 1/2` since P(X > 0) is 1/2. Moreover, I think rather than saying probability is 1, I should have said value of Z is 1.
> when z >0, the probability is 1 and when `z <= 0` then probability is 0.  
Hence, P(Z = 1) whenever X > 0, and P(X > 0) is 1/2 so, P(Z = 1) = 1/2. I totally agree with your opinion. Thanks for the correction. :-)
Agreed on you second comment too. I am working on some more tests as suggested by you. I will write the code after that.

Upabjojr replied:
There is a `Probability` class, it's meant for unevaluated probability expressions. I would suggest to start with that one.

That is, `P(..., X)` should return `Probability(..., X)` without evaluating the expression. In the future more specific algorithms may be added.

czgdp1807 replied:
@Upabjojr I have made some changes. I think this level of abstraction doesn't need more tests. 
I believe this partially solves the issue. However, more work is to be done for returning a random variable for P(A|X).

Upabjojr replied:
There is a potential problem that I see. Now `Probability` may be both a random and non-random expression, depending on the `given_condition` object. We need to devise some API change to support this.

Upabjojr replied:
I would suggest that SymPy objects be equipped with an `.is_random` property, defaulting to `False`.

Upabjojr replied:
But maybe this is for a future PR.

czgdp1807 replied:
@Upabjojr I agree, that this will create a problem. However, I think rather than giving, `is_random` attribute to all the sympy objects, instead, we can give, `is_random` to only `Probability` class and we can set the attribute to `True` according to the situation. This approach will require lesser and manageable changes and will not affect the code written earlier, as far as I can see.

Example:
```
>>> X = Normal('X', 0, 1)
>>> Z = P(X > 0, X > 0)
>>> Z.is_random
False
>>> Z = P(X > 0, X)
>>> Z.is_random
True
```

czgdp1807 committed:
Made modifications

czgdp1807 commented:
@Upabjojr I have made some changes using Bayes' theorem. 
I have a doubt that which probability space will `P(A|X)` follow? X's or the random variable in A? Because, we need `pspace` to create an object of `RandomSymbol`. Please clarify.

Upabjojr commented:
> I have a doubt that which probability space will `P(A|X)` follow?

The one of `X`.

Upabjojr commented:
> Because, we need `pspace` to create an object of `RandomSymbol`. Please clarify.

I have thought again, creating a `RandomSymbol` for `P(A, X)` is actually a bad idea. There is a class called `Probability` that should work fine.

Actually I would like `P` to point to `Probability` rather than `probability`, though this requires some deprecation procedure.

Upabjojr commented:
If `A` and `B` are independent, you should have `P(A|B) = P(A and B)/P(B) = P(A)`. The problem remains in the case the pspaces are dependent.

czgdp1807 commented:
@Upabjojr The code which I added has considered the case of independent random expressions.
Moreover, one more solution which I can suggest,
P(A|X) = P(A,X)/P(X)
Since, you suggested to point to `Probability` class rather than making a new random variable. So, I think the above expression can be useful provided, `Probability` supports joint random expressions. For continuous RVs,
PMFs can be replaced by PDFs.
P.S. Can you please provide a link to the details of deprecation procedure?

Upabjojr commented:
@czgdp1807 maybe you should start from some tests rather than rush to write the code.

> P.S. Can you please provide a link to the details of deprecation procedure?

Look up `SymPyDeprecationWarning` and its related decorator.

jksuom commented:
This is in wiki: https://github.com/sympy/sympy/wiki/Deprecating-policy

czgdp1807 commented:
Sure @Upabjojr. I will make some meaningful tests first. Infact, it will help me in making the issue more clear. Thanks for the suggestion. 
Thanks a lot @jksuom for the link to the policy.

czgdp1807 committed:
Abstraction of results

Modifications made to return
an object of Probability class
rather than the evaluated expression.
Tests have also been updated.

Upabjojr commented:
Did you do some more testing? Do you need to add the condition to `ProductDomain` only?

czgdp1807 committed:
Added a missing import

czgdp1807 commented:
@Upabjojr I am working on the tests. I will test them in the comments with their expected outputs shortly.

czgdp1807 commented:
@Upabjojr 
Please consider the following points, which I have developed while making tests.

- Let us say, `P(f(X), X)`, where, `f(X)` denotes events like, `X > 0`, `X < a`, etc. Consider a random variable `Z`, which represents, `Z = P(f(X), X)`. By observation, it is clear that `Z` can only attain two values, `1` and `0` because, for any given value of `X` either, `f(X)` will be satisfied or will not be satisfied. Hence, irrespective of probability space of `X`, `Z` will always follow Bernoulli Distribution, with parameter,` p = P(f(X))`.
- Let us consider the case of two random variables, `P(f(X, Y), X)`. Here, again, `f(X,  Y)` denotes events like, `X + Y > 0`, `X + Y < a`, etc. Again consider random variable, `R = P(f(X, Y), X)`. For any given value of `X`, the values of `Y` for which `f(X, Y)` is valid is given by the set, say, `V`. So, the value of `R` will be given by,  `P(Y lies in V)`. `f(X, Y)` can be solved for `Y` by using the `solve`. Similarly, P(f(X, Y), Y) can be interpreted.

For three or more variables, the analysis will be more rigorous.

Upabjojr commented:
>  `f(X, Y)` can be solved for `Y` by using the `solve`

Supposing `solve` can find the solution.

Upabjojr commented:
> For three or more variables, the analysis will be more rigorous.

I would recommend to do one step at a time. It's OK to return an unevaluated `Probability` object for now. Further simplifications can be thought of. Maybe you can open another PR with your ideas.

czgdp1807 commented:
So, I hope that this PR has solves the issue for now and no more work is needed. If more work is to be done then please mention like adding more tests, if needed.
I will open one more PR for the single variable case, i.e., P(f(X)|X), since it's pretty straight forward and moreover I believe it will be compatible with the future changes.

czgdp1807 committed:
added check for dependency of random variables

czgdp1807 commented:
@Upabjojr I have made changes for incorporating the following,

> If `A` and `B` are independent, you should have `P(A|B) = P(A and B)/P(B) = P(A)`. The problem remains in the case the pspaces are dependent.

Tests are also updated accordingly. Please see if it's good.

czgdp1807 committed:
modified tests

czgdp1807 committed:
modified tests

czgdp1807 committed:
updated tests for python

czgdp1807 committed:
removed extra line

czgdp1807 commented:
@Upabjojr All checks have passed. Any comments/reviews for more changes?

czgdp1807 commented:
@oscarbenjamin @Abdullahjavednesar Please update the labels.

czgdp1807 commented:
@Upabjojr All tests are passing for this PR. Please let me know if any other changes are to be made.
